### PR TITLE
Allow Auth credentials to be fetched via callback

### PR DESCRIPTION
**Pull request by JohnRbk**
For OIDC or Cognito authentication, auth tokens often expire and need to be re-generated. Using a callback follows the model also used by aws-mobile-appsync-sdk-js

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by rohandubal** [comment]
Hello @JohnRbk 

Thanks for the PR! This is a good use-case which we can add to the SDK. But the way the current changes are, this would be a breaking change for the consumers. Is there a non-breaking solution which you can think of and update the PR? 

I will discuss with the team about the possibility of introducing a breaking change in the meanwhile.

Thanks,
Rohan

**Comment by JohnRbk** [comment]
Hi Rohan, yes, I think there is a way to make this a non-breaking change. I do think it has trade-offs with respect to the programming model; even on the Android sdk I've seen examples where people have used semaphores to work around the lack of a callback, and that leads to overly complicated code. I think it would also make sense to align a solution across all platform SDKs.

Regardless, here is a way to make this a non-breaking change using a new protocol. Let me know what you think. (Maybe a fatalError() in the extension `getLatestAuthToken` would be better to avoid improper use).

In file AWSAppSyncAuthProvider.swift:

```swift
public protocol AWSOIDCAuthProviderAsync: AWSOIDCAuthProvider {
    func getLatestAuthToken(_ callback: @escaping (String) -> Void)
}
extension AWSOIDCAuthProviderAsync {
    public func getLatestAuthToken() -> String { return "" }
}
```

And in AWSAppSyncHTTPNetworkTransport.swift:

```swift
    private func sendRequestWithAuth(mutableRequest: NSMutableURLRequest, sendRequest: @escaping (URLRequest) -> Void ) {        
        switch self.authType {
            
        case .awsIAM:
            let signer:AWSSignatureV4Signer = AWSSignatureV4Signer(credentialsProvider: self.credentialsProvider, endpoint: self.endpoint)
            signer.interceptRequest(mutableRequest).continueWith { _ in
                return nil
                }.continueWith { _ in
                    sendRequest(mutableRequest as URLRequest)
            }
        case .apiKey:
            mutableRequest.setValue(self.apiKeyAuthProvider!.getAPIKey(), forHTTPHeaderField: "x-api-key")
            sendRequest( mutableRequest as URLRequest)
        case .oidcToken:
            
            if let authProvider = self.userPoolsAuthProvider as? AWSOIDCAuthProviderAsync {
                authProvider.getLatestAuthToken { token in
                    mutableRequest.setValue(token, forHTTPHeaderField: "authorization")
                    sendRequest( mutableRequest as URLRequest)
                }
            }
            else if let authProvider = self.userPoolsAuthProvider {
                mutableRequest.setValue(authProvider.getLatestAuthToken(), forHTTPHeaderField: "authorization")
                sendRequest( mutableRequest as URLRequest)
            }
            
        }
    }
```

**Comment by joelmbell** [comment]
I could really use this, any chance of it getting merged in soon?

**Comment by kvasukib** [comment]
@JohnRbk The PR has conflicts. I have touched the same pieces of code in my PR and that's the reason you may have conflicts. Can you resolve the conflicts and apply your non-breaking changes to the PR?

@joelmbell I'm waiting on @JohnRbk to resolve the conflicts. I will do a final review once the conflicts are resolved.

File: AWSAppSyncClient/AWSAppSyncAuthProvider.swift [hunk 1]
```diff
@@ -8,7 +8,7 @@ import Foundation
 // For using OIDC based authorization, this protocol needs to be implemented and passed to configuration object.
 public protocol AWSOIDCAuthProvider {
     /// The method should fetch the token and return it to the client for using in header request.
-    func getLatestAuthToken() -> String
+    func getLatestAuthToken(_ callback: @escaping (String) -> Void)
```
Comment (kvasukib) [inline AWSAppSyncClient/AWSAppSyncAuthProvider.swift]: Can you make this a non-breaking change?

**Comment by JohnRbk** [comment]
Hey guys, I re-baselined my project with the latest master & have applied my updates. Let me close out this PR and issue a new clean one. I also have some unit tests to add as well.

**Pull request closed without merging**